src/python/pants/BUILD

```diff
@@ -34,9 +34,12 @@ python_distribution(
     # consumers of pantsbuild.pants are using a compatible interpreter.
     # TODO(7344): the tuple syntax for ext_modules is deprecated. Use Extension once we support it.
     ext_modules=[('native_engine', {'sources': ['src/pants/dummy.c']})],
-  ).with_binaries(
-    pants='src/python/pants/bin:pants',
-  )
+  ),
+  entry_points={
+    'console_scripts': {
+      'pants': 'pants.bin.pants_loader:main',
```

Comment:
I do think it would be great to circle back and add rule support for parsing this field to figure out if it is a target address to support the old DRY `.with_binaries` provided. Relative address syntax (`:name`) is unambiguous, but otherwise we could just require `//explicit/absolute:form` and the `//` would make everything unambiguous. Perhaps add a TODO / tracking issue link if you agree.

Reply:
If we require the prefix of `//`, then I'm okay with it. As explained in https://github.com/pantsbuild/pants/issues/11808#issuecomment-811358327, I bias towards the simplicity of only having the module format. But requiring the unambiguous syntax does resolve half of my concern. 

The only remaining concern now is that we have two ways of doing something, which means more documentation etc. But that isn't too concerning:

> Blah blah...should be in the format `path.to.module` or `path.to.module:func`. You can also point to a `pex_binary` target to use its `entry_point` field by using either `:sibling_target` or `//path/to:target`.

What do you think @kaos?

Reply:
I'd like to keep it. It helps with DRYness, and also makes migration from `.with_binaries()` easier. 
The only downside I see is if it adds another difference in semantics to `entry_points` on `python_distribution` vs `setup_py()`.

Maybe deprecate `entry_points` on `setup_py()` and flag it as a reserved field (akin to packages and a few others), to get rid of the potential pitfalls.

Reply:
Sg on allowing pex_binary addresses if it's unambiguous by either `:tgt` or `//path/to:tgt` (`//path/to` is also valid, where the target's name == parent folder). If it does not start with either `//` or `:`, we assume it's a module.

> Maybe deprecate entry_points on setup_py() and flag it as a reserved field (akin to packages and a few others), to get rid of the potential pitfalls.

I don't think we need to do that - you can still use the old `setup_py()` way of doing it, only you'll miss out a) dependency inference, and b) being able to use a `pex_binary` address. By allowing `setup_py()`, we reduce possible churn and it allows people to dynamically compute the value with the plugin hook to generate `setup_py()` args: https://www.pantsbuild.org/docs/plugins-setup-py

Reply:
I opted for a check that if it starts with `:` or _contains_ any `/` we treat it as a target address, as this was the supported syntax for `.with_binaries()`, to ease migrations, but am open to switch to the stricter check if you prefer that.